File: dotnet/src/Envdoctor/Formatters/SarifFormatter.cs
using Envdoctor.Core;
using Envdoctor.Models;
using Envdoctor.Utils;

namespace Envdoctor.Formatters;

/// Render findings as SARIF 2.1.0 for GitHub code scanning.
public static class SarifFormatter
{
    private const string SarifSchema =
        "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json";

    /// Detectors whose default severity is `error`; everything else defaults
    /// to `warning`. Mirrors the reference `defaultLevelForDetector`.
    private static readonly string[] ErrorDetectors =
    {
        "missing",
        "undefined-in-source",
        "type-mismatch",
        "public-prefix",
    };

    public static string RenderSarif(IReadOnlyList<Finding> findings, string rootDir)
    {
        var results = findings.Select(f => FindingToSarif(f, rootDir)).ToList<object?>();

        return Json.Pretty(new JsonObject
        {
            { "$schema", SarifSchema },
            { "version", "2.1.0" },
            {
                "runs", new List<object?>
                {
                    new JsonObject
                    {
                        {
                            "tool", new JsonObject
                            {
                                {
                                    "driver", new JsonObject
                                    {
                                        { "name", "envdoctor" },
                                        { "informationUri", "https://github.com/arun-skg/envdoctor" },
                                        { "rules", RenderRules() },
                                    }
                                },
                            }
                        },
                        { "results", results },
                    },
                }
            },
        });
    }

    private static string SeverityToLevel(Severity severity) => severity switch
    {
        Severity.Error => "error",
        Severity.Warning => "warning",
        Severity.Info => "note",
        _ => "note",
    };

    private static object FindingToSarif(Finding f, string rootDir)
    {
        var locations = f.Locations.Select(o =>
        {
            var rel = Core.Discover.RelativeTo(rootDir, o.FilePath) ?? o.FilePath;
            var physical = new JsonObject
            {
                { "artifactLocation", new JsonObject { { "uri", rel.Replace('\\', '/') } } },
            };
            if (o.Line is { } line && line > 0)
                physical.Add("region", new JsonObject { { "startLine", line } });
            return (object)new JsonObject { { "physicalLocation", physical } };
        }).ToList();

        return new JsonObject
        {
            { "ruleId", f.RuleId },
            { "level", SeverityToLevel(f.Severity) },
            { "message", new JsonObject { { "text", $"{f.Variable}: {f.Message}" } } },
            { "locations", locations },
        };
    }

    /// The full detector catalog, matching the reference CLI (all rules
    /// appear, not only the ones with findings).
    private static List<object> RenderRules() =>
        Audit.AllDetectors()
            .Select(d =>
            {
                var level = ErrorDetectors.Contains(d.Id) ? "error" : "warning";
                return (object)new JsonObject
                {
                    { "id", d.Id },
                    { "name", d.Name },
                    { "shortDescription", new JsonObject { { "text", d.Description } } },
                    { "defaultConfiguration", new JsonObject { { "level", level } } },
                };
            })
            .ToList();
}

File: dart/lib/src/formatters/sarif_formatter.dart
import '../core/audit.dart';
import '../models/finding.dart';
import '../utils/json.dart';
import '../utils/paths.dart';

const String sarifSchema =
    'https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json';

/// Detectors whose default severity is `error`; everything else defaults
/// to `warning`. Mirrors the reference `defaultLevelForDetector`.
const List<String> errorDetectors = [
  'missing',
  'undefined-in-source',
  'type-mismatch',
  'public-prefix',
];

/// Render findings as SARIF 2.1.0 for GitHub code scanning.
String renderSarif(List<Finding> findings, String rootDir) {
  final results = findings.map((f) => _findingToSarif(f, rootDir)).toList();

  return jsonPretty(JsonObject()
    ..add('\$schema', sarifSchema)
    ..add('version', '2.1.0')
    ..add('runs', [
      JsonObject()
        ..add('tool', JsonObject()
          ..add('driver', JsonObject()
            ..add('name', 'envdoctor')
            ..add('informationUri', 'https://github.com/arun-skg/envdoctor')
            ..add('rules', _renderRules())))
        ..add('results', results),
    ]));
}

String _severityToLevel(Severity severity) => switch (severity) {
      Severity.error => 'error',
      Severity.warning => 'warning',
      Severity.info => 'note',
    };

Object _findingToSarif(Finding f, String rootDir) {
  final locations = f.locations.map((o) {
    final rel = relativeTo(rootDir, o.filePath) ?? o.filePath;
    final physical = JsonObject()
      ..add('artifactLocation', JsonObject()
        ..add('uri', normalizePath(rel)));
    final line = o.line;
    if (line != null && line > 0) {
      physical.add('region', JsonObject()..add('startLine', line));
    }
    return JsonObject()..add('physicalLocation', physical);
  }).toList();

  return JsonObject()
    ..add('ruleId', f.ruleId)
    ..add('level', _severityToLevel(f.severity))
    ..add('message', JsonObject()..add('text', '${f.variable}: ${f.message}'))
    ..add('locations', locations);
}

/// The full detector catalog, matching the reference CLI (all rules appear,
/// not only the ones with findings).
List<Object> _renderRules() => allDetectors().map((d) {
      final level = errorDetectors.contains(d.id) ? 'error' : 'warning';
      return JsonObject()
        ..add('id', d.id)
        ..add('name', d.name)
        ..add('shortDescription', JsonObject()..add('text', d.description))
        ..add('defaultConfiguration', JsonObject()..add('level', level));
    }).toList();
